expression-language.md

```diff
@@ -0,0 +1,375 @@
+| `UPPER(x): String -> String`                               | Returns `x` in uppercase                                                                                                                                                                                                   |
+| `TRIM(x): String -> String`                                | Returns `x` with leading and trailing trimmed whitespaces                                                                                                                                                                  |
+| `LEFT(x, y): String x Integer -> String`                   | Returns a new string with the first `y` characters of `x`, or returns `x` if `LENGTH(x) <= y`. Returns `x` if `y < 0` and raise an error                                                                                   |
+| `RIGHT(x, y): String x Integer -> String`                  | Returns a new string with the last `y` characters of `x` or returns `x` if `LENGTH(x) <= y`. Returns `x` if `y < 0` and raise an error                                                                                     |
+| `SUBSTRING(x, b, e): String x Integer x Integer -> String` | Returns the substring of `x` starting from index `b` (included) and ending with index `e` (excluded). Returns `SUBSTRING(x, b, LENGTH(x))` if `LENGTH(x) < e`. Returns `x` and raise an error if `b < 0 OR e < 0 OR e < b` |
+
+### 3.6. Evaluation of the expression
+
+Operators MUST be evaluated in order, where the parenthesized expressions have the highest priority over all the other
+operators.
+
+A CESQL expression, when evaluated, MUST always return a Boolean `true` if the evaluation of the predicate described by
+the expression matches the input, otherwise `false`.
+
+An evaluation might return an error together with the return value, which the evaluator MUST notify to the user.
+
+## 4. Examples
+
+_CloudEvent including a subject_
+
+```
+EXISTS subject
+```
+
+_CloudEvent including the extension 'firstname' with value 'Francesco'_
+
+```
+firstname = 'Francesco'
+```
+
+_CloudEvent including the extension 'firstname' with value 'Francesco' or the subject with value 'Francesco'_
+
+```
+firstname = 'Francesco' OR subject = 'Francesco'
+```
+
+_CloudEvent including the extension 'firstname' with value 'Francesco' and extension 'lastname' with value 'Guardiani',
+or the subject with value 'Francesco Guardiani'_
+
+```
+(firstname = 'Francesco' AND lastname = 'Guardiani') OR subject = 'Francesco Guardiani'
+```
+
+_CloudEvent including the extension 'sequence' with numeric value 10_
+
+```
+INT(sequence) = 10
+```
+
+_CloudEvent including the extension 'hop' and 'ttl', where 'hop' is smaller than 'ttl'_
+
+```
+INT(hop) < INT(ttl)
```

Comment:
< is only allowed for numerical types and therefore the type casts can also be inferred